{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HOMEWORK 2 - PROGRAMMING PORTION\n",
    "#### Genevieve Mortensen 2023-03-20\n",
    "----------------------------------------\n",
    "\n",
    "The purpose of this script is to compare cancer type predictions produced by Logistic Regression, Naive Bayes, and LDA classification methods based on gene mutation data.\n",
    "\n",
    "The data is obtained here: https://www.cbioportal.org/study/summary?id=brca_tcga_pub2015\n",
    "\n",
    "## Q2.1) Extract data\n",
    "\n",
    "Here, we need to subset the mutation data for patients with invasive ductal and lobular carcinoma, which are the two most common types of breast cancer, respectively. This is the case where a patient's cancer (carcinoma) begins in the breast's milk transport duct or production gland (lobule) and then spreads to other tissues in the body.\n",
    "\n",
    "                Based on the biological context, what does this mean for our classifier?\n",
    "\n",
    "We want to classify the mutations characterizing these *two* types of cancer based on their distributions in the data. We therefore will create a __*binary classifier*__ using the three different methods aforementioned.\n",
    "\n",
    "### Import Libraries\n",
    "\n",
    "There are a LOT of libraries that we will need for this work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import libraries\n",
    "\n",
    "#Basic data handling libraries\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt, seaborn as sns\n",
    "%matplotlib inline\n",
    "sns.set(rc={'axes.facecolor':'lightblue'})\n",
    "\n",
    "#Scikitlearn libraries\n",
    "import sklearn\n",
    "from sklearn import metrics\n",
    "from sklearn import model_selection\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics import classification_report, precision_recall_curve\n",
    "from sklearn.model_selection import StratifiedKFold, train_test_split, RepeatedStratifiedKFold, GridSearchCV, cross_val_score\n",
    "from sklearn.linear_model import LinearRegression, LogisticRegression\n",
    "from sklearn.feature_selection import RFE\n",
    "from sklearn.impute import KNNImputer\n",
    "from sklearn.datasets import make_blobs\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import make_scorer, accuracy_score, precision_score, recall_score, f1_score\n",
    "from sklearn.naive_bayes import GaussianNB, BernoulliNB\n",
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis\n",
    "\n",
    "\n",
    "# Statistics import\n",
    "from statistics import mean, stdev\n",
    "\n",
    "\n",
    "# Importing statsmodels\n",
    "import statsmodels.api as sm\n",
    "\n",
    "# Importing 'variance_inflation_factor'\n",
    "from statsmodels.stats.outliers_influence import variance_inflation_factor\n",
    "\n",
    "# Importing warnings\n",
    "# ------------------\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# Removing the limit for displaying rows and columns\n",
    "# --------------------------------------------------\n",
    "# pd.set_option('display.max_rows', None)\n",
    "# pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Wrangle Data\n",
    "\n",
    "The TA gave us most of the solution to this question, so we will use it here. There are many ways to pull out the relevant data, but basically we just need to have an sparse __n x d__ matrix, with our categories for classification also included in the dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(56933, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Gene</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTGER3</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000356595</td>\n",
       "      <td>ENST00000356595</td>\n",
       "      <td>ENSG00000050628</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLG</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000368799</td>\n",
       "      <td>ENST00000368799</td>\n",
       "      <td>ENSG00000143631</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GPR52</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000367685</td>\n",
       "      <td>ENST00000367685</td>\n",
       "      <td>ENSG00000203737</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SLC35F3</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000366618</td>\n",
       "      <td>ENST00000366618</td>\n",
       "      <td>ENSG00000183780</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR2T3</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000359594</td>\n",
       "      <td>ENST00000359594</td>\n",
       "      <td>ENSG00000196539</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62015</th>\n",
       "      <td>CCDC97</td>\n",
       "      <td>TCGA-BH-A42T</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000269967</td>\n",
       "      <td>ENST00000269967</td>\n",
       "      <td>ENSG00000142039</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62016</th>\n",
       "      <td>SMG9</td>\n",
       "      <td>TCGA-BH-A42T</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000270066</td>\n",
       "      <td>ENST00000270066</td>\n",
       "      <td>ENSG00000105771</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62017</th>\n",
       "      <td>ZNF324</td>\n",
       "      <td>TCGA-BH-A42T</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000536459</td>\n",
       "      <td>ENST00000536459</td>\n",
       "      <td>ENSG00000083812</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62018</th>\n",
       "      <td>PLTP</td>\n",
       "      <td>TCGA-BH-A42T</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000477313</td>\n",
       "      <td>ENST00000477313</td>\n",
       "      <td>ENSG00000100979</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70013</th>\n",
       "      <td>CABYR</td>\n",
       "      <td>TCGA-BH-A0HP</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000399496</td>\n",
       "      <td>ENST00000399496</td>\n",
       "      <td>ENSG00000154040</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56933 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Hugo_Symbol    PATIENT_ID Variant_Classification Variant_Type  \\\n",
       "0          PTGER3  TCGA-B6-A0IG      Missense_Mutation          SNP   \n",
       "1             FLG  TCGA-B6-A0IG      Missense_Mutation          SNP   \n",
       "2           GPR52  TCGA-B6-A0IG                 Silent          SNP   \n",
       "3         SLC35F3  TCGA-B6-A0IG                 Silent          SNP   \n",
       "4           OR2T3  TCGA-B6-A0IG      Missense_Mutation          SNP   \n",
       "...           ...           ...                    ...          ...   \n",
       "62015      CCDC97  TCGA-BH-A42T      Missense_Mutation          SNP   \n",
       "62016        SMG9  TCGA-BH-A42T      Missense_Mutation          SNP   \n",
       "62017      ZNF324  TCGA-BH-A42T      Missense_Mutation          SNP   \n",
       "62018        PLTP  TCGA-BH-A42T      Missense_Mutation          SNP   \n",
       "70013       CABYR  TCGA-BH-A0HP            Splice_Site          SNP   \n",
       "\n",
       "         Transcript_ID          Feature             Gene  \\\n",
       "0      ENST00000356595  ENST00000356595  ENSG00000050628   \n",
       "1      ENST00000368799  ENST00000368799  ENSG00000143631   \n",
       "2      ENST00000367685  ENST00000367685  ENSG00000203737   \n",
       "3      ENST00000366618  ENST00000366618  ENSG00000183780   \n",
       "4      ENST00000359594  ENST00000359594  ENSG00000196539   \n",
       "...                ...              ...              ...   \n",
       "62015  ENST00000269967  ENST00000269967  ENSG00000142039   \n",
       "62016  ENST00000270066  ENST00000270066  ENSG00000105771   \n",
       "62017  ENST00000536459  ENST00000536459  ENSG00000083812   \n",
       "62018  ENST00000477313  ENST00000477313  ENSG00000100979   \n",
       "70013  ENST00000399496  ENST00000399496  ENSG00000154040   \n",
       "\n",
       "                     CANCER_TYPE               CANCER_TYPE_DETAILED  \n",
       "0      Invasive Breast Carcinoma   Breast Invasive Ductal Carcinoma  \n",
       "1      Invasive Breast Carcinoma   Breast Invasive Ductal Carcinoma  \n",
       "2      Invasive Breast Carcinoma   Breast Invasive Ductal Carcinoma  \n",
       "3      Invasive Breast Carcinoma   Breast Invasive Ductal Carcinoma  \n",
       "4      Invasive Breast Carcinoma   Breast Invasive Ductal Carcinoma  \n",
       "...                          ...                                ...  \n",
       "62015  Invasive Breast Carcinoma  Breast Invasive Lobular Carcinoma  \n",
       "62016  Invasive Breast Carcinoma  Breast Invasive Lobular Carcinoma  \n",
       "62017  Invasive Breast Carcinoma  Breast Invasive Lobular Carcinoma  \n",
       "62018  Invasive Breast Carcinoma  Breast Invasive Lobular Carcinoma  \n",
       "70013  Invasive Breast Carcinoma  Breast Invasive Lobular Carcinoma  \n",
       "\n",
       "[56933 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## TA PROVIDED SOLUTION ##\n",
    "# Reading \"clinicalsample\" dataset\n",
    "df_clinicalsample = pd.read_csv(r\"./brca_tcga_pub2015/brca_tcga_pub2015/data_clinical_sample.txt\", sep=\"\\t\", skiprows=4)\n",
    "df_clinicalsample.replace('[Not Available]', np.nan, inplace=True)\n",
    "# df_clinicalpatient['Tumor_Sample_Barcode'] = df_clinicalpatient['PATIENT_ID']\n",
    "\n",
    "# Reading \"data_mutations.txt\" dataset\n",
    "df_mutations = pd.read_csv(r\"./brca_tcga_pub2015/brca_tcga_pub2015/data_mutations.txt\", sep=\"\\t\")\n",
    "\n",
    "# Slicing Tumor Barcodes\n",
    "df_mutations['PATIENT_ID'] = df_mutations['Tumor_Sample_Barcode'].apply(lambda x: x[:-3])\n",
    "\n",
    "\n",
    "# Merging the values\n",
    "# ------------------\n",
    "df = pd.merge(df_mutations, df_clinicalsample, on='PATIENT_ID', how='left')\n",
    "df.head()\n",
    "\n",
    "# Filtering Out the important columns in the Dataframe, that we are going to use\n",
    "# ------------------------------------------------------------------------------\n",
    "cancer = df[['Hugo_Symbol','PATIENT_ID','Variant_Classification',\n",
    "             'Variant_Type','Transcript_ID','Feature','Gene',\n",
    "             'CANCER_TYPE','CANCER_TYPE_DETAILED',]]\n",
    "\n",
    "\n",
    "# Filtering out Patients with the below 2 types of Carcinoma's\n",
    "# --> Breast Invasive Ductal Carcinoma - \"Infiltrating Ductal Carcinoma\"\n",
    "# --> Breast Invasice Lobular Carcinoma - \"Infiltrating Lobular Carcinoma\"\n",
    "#\n",
    "cancer = cancer.loc[(cancer['CANCER_TYPE_DETAILED']=='Breast Invasive Lobular Carcinoma') | (cancer['CANCER_TYPE_DETAILED']=='Breast Invasive Ductal Carcinoma')]\n",
    "\n",
    "# Target Variable:\n",
    "df_target = cancer[['PATIENT_ID','CANCER_TYPE_DETAILED']].drop_duplicates()\n",
    "\n",
    "print(cancer.shape)\n",
    "display(cancer)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Basically, the hard part was done for us. Thanks, Nitanshu!\n",
    "\n",
    "### Preprocessing\n",
    "\n",
    "Let's move on to creating the matrix. We can use One Hot Encoding to turn our mutations into binary categorical variables. This will give us a sparse matrix correlating to the number of times the mutation appears in a given patient. \n",
    "We will also need to retain the cancer type details for our binary classification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AADACL2</th>\n",
       "      <th>AADACL4</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62015</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62016</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62017</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62018</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70013</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56933 rows × 15302 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       A2M  A2ML1  A4GALT  A4GNT  AAAS  AACS  AADAC  AADACL2  AADACL4  AADAT  \\\n",
       "0        0      0       0      0     0     0      0        0        0      0   \n",
       "1        0      0       0      0     0     0      0        0        0      0   \n",
       "2        0      0       0      0     0     0      0        0        0      0   \n",
       "3        0      0       0      0     0     0      0        0        0      0   \n",
       "4        0      0       0      0     0     0      0        0        0      0   \n",
       "...    ...    ...     ...    ...   ...   ...    ...      ...      ...    ...   \n",
       "62015    0      0       0      0     0     0      0        0        0      0   \n",
       "62016    0      0       0      0     0     0      0        0        0      0   \n",
       "62017    0      0       0      0     0     0      0        0        0      0   \n",
       "62018    0      0       0      0     0     0      0        0        0      0   \n",
       "70013    0      0       0      0     0     0      0        0        0      0   \n",
       "\n",
       "       ...  ZWINT  ZXDA  ZXDB  ZXDC  ZYG11A  ZYG11B  ZYX  ZZEF1  ZZZ3  \\\n",
       "0      ...      0     0     0     0       0       0    0      0     0   \n",
       "1      ...      0     0     0     0       0       0    0      0     0   \n",
       "2      ...      0     0     0     0       0       0    0      0     0   \n",
       "3      ...      0     0     0     0       0       0    0      0     0   \n",
       "4      ...      0     0     0     0       0       0    0      0     0   \n",
       "...    ...    ...   ...   ...   ...     ...     ...  ...    ...   ...   \n",
       "62015  ...      0     0     0     0       0       0    0      0     0   \n",
       "62016  ...      0     0     0     0       0       0    0      0     0   \n",
       "62017  ...      0     0     0     0       0       0    0      0     0   \n",
       "62018  ...      0     0     0     0       0       0    0      0     0   \n",
       "70013  ...      0     0     0     0       0       0    0      0     0   \n",
       "\n",
       "                    CANCER_TYPE_DETAILED  \n",
       "0       Breast Invasive Ductal Carcinoma  \n",
       "1       Breast Invasive Ductal Carcinoma  \n",
       "2       Breast Invasive Ductal Carcinoma  \n",
       "3       Breast Invasive Ductal Carcinoma  \n",
       "4       Breast Invasive Ductal Carcinoma  \n",
       "...                                  ...  \n",
       "62015  Breast Invasive Lobular Carcinoma  \n",
       "62016  Breast Invasive Lobular Carcinoma  \n",
       "62017  Breast Invasive Lobular Carcinoma  \n",
       "62018  Breast Invasive Lobular Carcinoma  \n",
       "70013  Breast Invasive Lobular Carcinoma  \n",
       "\n",
       "[56933 rows x 15302 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Get encoding for mutations\n",
    "dummies = pd.get_dummies(cancer[\"Hugo_Symbol\"], drop_first=True)\n",
    "#concatenate cancer type\n",
    "dummies = pd.concat([dummies, cancer[\"CANCER_TYPE_DETAILED\"]], axis=1)\n",
    "display(dummies)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's fix our matrix such that our mutations are grouped by the patients and we retain cancer type details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AADACL2</th>\n",
       "      <th>AADACL4</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SD</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SE</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SF</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SI</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SJ</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-LL-A5YP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-LQ-A4E4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-MS-A51U</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OL-A66I</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OL-A66K</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>617 rows × 15302 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              A2M  A2ML1  A4GALT  A4GNT  AAAS  AACS  AADAC  AADACL2  AADACL4  \\\n",
       "PATIENT_ID                                                                     \n",
       "TCGA-A1-A0SD    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-A1-A0SE    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-A1-A0SF    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-A1-A0SI    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-A1-A0SJ    0      0       0      0     0     0      0        0        0   \n",
       "...           ...    ...     ...    ...   ...   ...    ...      ...      ...   \n",
       "TCGA-LL-A5YP    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-LQ-A4E4    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-MS-A51U    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-OL-A66I    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-OL-A66K    0      0       0      0     0     0      0        0        0   \n",
       "\n",
       "              AADAT  ...  ZWINT  ZXDA  ZXDB  ZXDC  ZYG11A  ZYG11B  ZYX  ZZEF1  \\\n",
       "PATIENT_ID           ...                                                        \n",
       "TCGA-A1-A0SD      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-A1-A0SE      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-A1-A0SF      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-A1-A0SI      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-A1-A0SJ      0  ...      0     0     0     0       0       0    0      0   \n",
       "...             ...  ...    ...   ...   ...   ...     ...     ...  ...    ...   \n",
       "TCGA-LL-A5YP      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-LQ-A4E4      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-MS-A51U      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-OL-A66I      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-OL-A66K      0  ...      0     0     0     0       0       0    0      0   \n",
       "\n",
       "              ZZZ3               CANCER_TYPE_DETAILED  \n",
       "PATIENT_ID                                             \n",
       "TCGA-A1-A0SD     0   Breast Invasive Ductal Carcinoma  \n",
       "TCGA-A1-A0SE     0  Breast Invasive Lobular Carcinoma  \n",
       "TCGA-A1-A0SF     0   Breast Invasive Ductal Carcinoma  \n",
       "TCGA-A1-A0SI     0   Breast Invasive Ductal Carcinoma  \n",
       "TCGA-A1-A0SJ     0   Breast Invasive Ductal Carcinoma  \n",
       "...            ...                                ...  \n",
       "TCGA-LL-A5YP     0   Breast Invasive Ductal Carcinoma  \n",
       "TCGA-LQ-A4E4     0  Breast Invasive Lobular Carcinoma  \n",
       "TCGA-MS-A51U     0  Breast Invasive Lobular Carcinoma  \n",
       "TCGA-OL-A66I     0   Breast Invasive Ductal Carcinoma  \n",
       "TCGA-OL-A66K     0  Breast Invasive Lobular Carcinoma  \n",
       "\n",
       "[617 rows x 15302 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#concatenate full data frame for patient IDs\n",
    "df = pd.concat([cancer, dummies], axis = 1)\n",
    "#Group sparse matrix by patient IDs\n",
    "sparse_df = df.groupby([\"PATIENT_ID\"])[dummies.columns].max()\n",
    "sparse_df = sparse_df.loc[:,~sparse_df.columns.duplicated()]\n",
    "display(sparse_df)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "                    Cool! Now what?\n",
    "\n",
    "Now we have our sparse matrix. But, we need to prepare our data some more for our model training."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AADACL2</th>\n",
       "      <th>AADACL4</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A1-A0SD</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A1-A0SE</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A1-A0SF</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A1-A0SI</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A1-A0SJ</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612</th>\n",
       "      <td>TCGA-LL-A5YP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>TCGA-LQ-A4E4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>TCGA-MS-A51U</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>TCGA-OL-A66I</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>TCGA-OL-A66K</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>617 rows × 15303 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       PATIENT_ID  A2M  A2ML1  A4GALT  A4GNT  AAAS  AACS  AADAC  AADACL2  \\\n",
       "0    TCGA-A1-A0SD    0      0       0      0     0     0      0        0   \n",
       "1    TCGA-A1-A0SE    0      0       0      0     0     0      0        0   \n",
       "2    TCGA-A1-A0SF    0      0       0      0     0     0      0        0   \n",
       "3    TCGA-A1-A0SI    0      0       0      0     0     0      0        0   \n",
       "4    TCGA-A1-A0SJ    0      0       0      0     0     0      0        0   \n",
       "..            ...  ...    ...     ...    ...   ...   ...    ...      ...   \n",
       "612  TCGA-LL-A5YP    0      0       0      0     0     0      0        0   \n",
       "613  TCGA-LQ-A4E4    0      0       0      0     0     0      0        0   \n",
       "614  TCGA-MS-A51U    0      0       0      0     0     0      0        0   \n",
       "615  TCGA-OL-A66I    0      0       0      0     0     0      0        0   \n",
       "616  TCGA-OL-A66K    0      0       0      0     0     0      0        0   \n",
       "\n",
       "     AADACL4  ...  ZWINT  ZXDA  ZXDB  ZXDC  ZYG11A  ZYG11B  ZYX  ZZEF1  ZZZ3  \\\n",
       "0          0  ...      0     0     0     0       0       0    0      0     0   \n",
       "1          0  ...      0     0     0     0       0       0    0      0     0   \n",
       "2          0  ...      0     0     0     0       0       0    0      0     0   \n",
       "3          0  ...      0     0     0     0       0       0    0      0     0   \n",
       "4          0  ...      0     0     0     0       0       0    0      0     0   \n",
       "..       ...  ...    ...   ...   ...   ...     ...     ...  ...    ...   ...   \n",
       "612        0  ...      0     0     0     0       0       0    0      0     0   \n",
       "613        0  ...      0     0     0     0       0       0    0      0     0   \n",
       "614        0  ...      0     0     0     0       0       0    0      0     0   \n",
       "615        0  ...      0     0     0     0       0       0    0      0     0   \n",
       "616        0  ...      0     0     0     0       0       0    0      0     0   \n",
       "\n",
       "                  CANCER_TYPE_DETAILED  \n",
       "0     Breast Invasive Ductal Carcinoma  \n",
       "1    Breast Invasive Lobular Carcinoma  \n",
       "2     Breast Invasive Ductal Carcinoma  \n",
       "3     Breast Invasive Ductal Carcinoma  \n",
       "4     Breast Invasive Ductal Carcinoma  \n",
       "..                                 ...  \n",
       "612   Breast Invasive Ductal Carcinoma  \n",
       "613  Breast Invasive Lobular Carcinoma  \n",
       "614  Breast Invasive Lobular Carcinoma  \n",
       "615   Breast Invasive Ductal Carcinoma  \n",
       "616  Breast Invasive Lobular Carcinoma  \n",
       "\n",
       "[617 rows x 15303 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Fix up the dataframe\n",
    "df_logir = sparse_df.copy()\n",
    "df_logir.reset_index(level=0,inplace=True)\n",
    "display(df_logir)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "                  ~ Marvelous ~ Can we start classifying yet?\n",
    "\n",
    "No! Let's drop off these categorical data so that we can define our training data, which must be numeric as our independent variable, and then our ground truth labels, or our dependent variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X and y shape are:  (617, 15301) (617,)\n"
     ]
    }
   ],
   "source": [
    "#Independent variable x\n",
    "x = df_logir.drop('CANCER_TYPE_DETAILED', axis = 1)\n",
    "x = x.drop('PATIENT_ID',axis=1)\n",
    "#Dependent variable y\n",
    "y = df_logir['CANCER_TYPE_DETAILED']\n",
    "print(\"X and y shape are: \", x.shape, y.shape)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great, now we have a matrix X which houses our data and a vector Y which houses our ground truth labels. Given some vector of data in X we will get one label in Y. In other words, given some mutations for a single patient, we will get one class of cancer.\n",
    "\n",
    "## 2.2) Cross-validate classification models\n",
    "\n",
    "Our classification models will be trained on our pre-processed data and then we will use them to make some prediction based on that training. First we will establish some variables which will be used throughout the model testing and then we will decipher the specifics of each method. For all methods testing, we are doing supervised learning."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this step we scale the data to a normal distribution - that is, we transform and fit the data to a Gaussian distribution centered around a 0 mean."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scale input features\n",
    "scaler = StandardScaler()\n",
    "x_scaled = scaler.fit_transform(x)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we establish a some scoring parameters. \n",
    "- Accuracy measures the total correctness of a model. It is the number of correct predictions a model made throughout the entire dataset. This doesn't work if the data is not class-balanced, meaning we have the same number of labels of each class in our training data.\n",
    "- Precision measures how many labels for a class were correct. It is the number of correct labels for a class divided by the total number of labels assigned for a single class.\n",
    "- Recall measures how many samples for a class label were correctly labeled. It is the number of correct labels for a single class divided by the sum of the labels given for one class which were correct and the other class which were incorrect.\n",
    "- F1 score is the harmonic mean of precision and recall. It is the number of correct labels for a class divided by the sum of the correct labels for a class and half of the total number of incorrect labels for either class. This gives a better description of the quality of the classifier.\n",
    "\n",
    "For all of these, the better classifier is one where the value of the mteric is closer to 1. However, precision and recall tend to fluctuate harmonically between them, and accuracy wouldn't be reliable since we don't know if our data is class-balanced. \n",
    "\n",
    "Given this, we will use F1 score to evaluate our data. Thankfully, we don't need to calculate this on our own, but we could easily do this given our precision and recall scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "#scoring scheme for our model evaluation metrics\n",
    "scoring = {'accuracy' : make_scorer(accuracy_score), \n",
    "           'precision' : make_scorer(precision_score),\n",
    "           'recall' : make_scorer(recall_score), \n",
    "           'f1_score' : make_scorer(f1_score)}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we need to split our data into k-folds where k = 5, since we are performing 5-fold cross-validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining the Stratified K-Fold Cross Validation Object \n",
    "skf = KFold(n_splits=5, shuffle=True, random_state=42)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Awesome, we are done with the preparation. Now we can move on to classifying.\n",
    "\n",
    "\n",
    "                                 Classify me, Captain!"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### __Logistic Regression__\n",
    "\n",
    "Logistic regression classification is where we define some threshold on a logistic curve and fit our data to it.\n",
    "The logistic curve gives us a threshold at the inflection point and extrema at the asymptotes. For binary classification, this will work decently well- if a datum is above the threhold, it is given one class, below it, and it is given another."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining the Logistic Regression Model Object\n",
    "logir_ovr = LogisticRegression(random_state=0, \n",
    "                               multi_class='ovr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define lists to hold Model Evaluation Metrics\n",
    "logit_accuracies_1 = []\n",
    "logit_precision_1 = []\n",
    "logit_recall_1 = []\n",
    "logit_f1scores_1 = []"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each index in our data, we have a training set and a testing set. We first establish the indeces for each iteration of our data, or fold, which is 5-fold. 4 of the folds are used to train and one is used to test. Then, we fit our model from our training data, which essentially just changes the logistic function so that it characterizes our data. Then, we predict a label using the testing fold. Lastly, we calculate the different metrics which tell us how well the model predicted a label using the testing fold of the data.\n",
    "\n",
    "In the next iteration, we test on a different fold of the data, and train with the other remaining 4 folds. After 5 iterations of this, we have trained our model on all 5 folds of the data and tested it on all 5 folds.\n",
    "The 'macro' parameter in our scoring function schemes averages the class-wise F1 scores, so this is how we obtain an F1 score for each iteration. We will store all of these."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform 5-fold cross validation\n",
    "for train_index, test_index in skf.split(x, y):\n",
    "    \n",
    "    #assign current training data from scaled independent variable and dependent variable\n",
    "    x_train_fold, x_test_fold = x_scaled[train_index], x_scaled[test_index]\n",
    "    y_train_fold, y_test_fold = y[train_index], y[test_index]\n",
    "    \n",
    "    # Fit the model from our training data\n",
    "    logir_ovr.fit(x_train_fold, y_train_fold)\n",
    "    \n",
    "    # use the model to predict the class based on \n",
    "    y_test_fold_pred = logir_ovr.predict(x_test_fold)\n",
    "    \n",
    "    #Store results\n",
    "    logit_accuracies_1.append(logir_ovr.score(x_test_fold, y_test_fold))\n",
    "    logit_f1scores_1.append(metrics.f1_score(y_test_fold, y_test_fold_pred, average='macro'))\n",
    "    logit_precision_1.append(metrics.precision_score(y_test_fold, y_test_fold_pred, average='macro'))\n",
    "    logit_recall_1.append(metrics.recall_score(y_test_fold, y_test_fold_pred, average='macro'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just to be thorough, we show the precision and recall metrics as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.3951612903225806, 0.39919354838709675, 0.3983739837398374, 0.4146341463414634, 0.3780487804878049]\n",
      "[0.5, 0.5, 0.5, 0.5, 0.5]\n",
      "[0.44144144144144143, 0.4439461883408072, 0.4434389140271493, 0.45333333333333337, 0.4305555555555556]\n"
     ]
    }
   ],
   "source": [
    "print(logit_precision_1)\n",
    "print(logit_recall_1)\n",
    "print(logit_f1scores_1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### __Naive Bayes__\n",
    "\n",
    "Naive Bayes classifier assumes strong independence between samples and computes conditional probability of a label given the training data. It is known to be a very good classifier because each feature contributes independently to the probability of data belonging to a class. The posterior probability of class given the data is equivalent to the probability of that class appearing in the data multiplied by the probability of the data belonging to a class divided by the probability of that data existing. This will give us probabilities for each class and the higher probability wins, thus we assign the data the corresponding class label, following Bayes Theorem. The primary advantage is that class boundaries are variable, not just linear. We would expect this classifier to be the best one.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the object for Naive Bayes - Bernoulli NB is for boolean values, perfect for binary classification\n",
    "naive_bayes = BernoulliNB()\n",
    "\n",
    "#Create lists to hold metrics\n",
    "naive_accuracies_1 = []\n",
    "naive_precision_1 = []\n",
    "naive_recall_1 = []\n",
    "naive_f1scores_1 = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "for train_index, test_index in skf.split(x, y):  # 5 Times\n",
    "    \n",
    "    x_train_fold, x_test_fold = x_scaled[train_index], x_scaled[test_index]\n",
    "    y_train_fold, y_test_fold = y[train_index], y[test_index]\n",
    "    \n",
    "    naive_bayes.fit(x_train_fold, y_train_fold)\n",
    "    \n",
    "    # Predicting from predictor varialbes based on the model.\n",
    "    y_test_fold_pred = naive_bayes.predict(x_test_fold)\n",
    "    \n",
    "    #Store results\n",
    "    naive_accuracies_1.append(naive_bayes.score(x_test_fold, y_test_fold))\n",
    "    naive_f1scores_1.append(metrics.f1_score(y_test_fold, y_test_fold_pred, average='macro'))\n",
    "    naive_precision_1.append(metrics.precision_score(y_test_fold, y_test_fold_pred, average='macro'))\n",
    "    naive_recall_1.append(metrics.recall_score(y_test_fold, y_test_fold_pred, average='macro'))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.5649717514124294, 0.6033613445378152, 0.510233918128655, 0.47554347826086957, 0.5031512605042017]\n",
      "[0.5180533751962323, 0.5248484848484849, 0.5042857142857143, 0.48949579831932777, 0.5005376344086022]\n",
      "[0.4976851851851851, 0.5070336391437309, 0.48806881243063266, 0.47227077705386944, 0.4539400665926748]\n"
     ]
    }
   ],
   "source": [
    "print(naive_precision_1)\n",
    "print(naive_recall_1)\n",
    "print(naive_f1scores_1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### __LDA__\n",
    "\n",
    "This dimension-reducing technique uses Bayes Theorem to assign classes in a similar way, but draws linear boundaries between data instead of variable boundaries, so our boundaries depend on the total probability of a set of data corresponding to a class. It is shorter to evaluate and is expected to perform better than logistic regression, but not as well as naive bayes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the object for LDA\n",
    "lda = LinearDiscriminantAnalysis()\n",
    "\n",
    "# Define the object K-Fold Cross Validation\n",
    "lkfold = KFold(n_splits = 5,shuffle=True, random_state=42)\n",
    "\n",
    "#Create lists to hold metrics\n",
    "lda_accuracies_1 = []\n",
    "lda_precision_1 = []\n",
    "lda_recall_1 = []\n",
    "lda_f1scores_1 = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "for train_index, test_index in skf.split(x, y):  # 5 Times\n",
    "    \n",
    "    x_train_fold, x_test_fold = x_scaled[train_index], x_scaled[test_index]\n",
    "    y_train_fold, y_test_fold = y[train_index], y[test_index]\n",
    "    \n",
    "    lda.fit(x_train_fold, y_train_fold)\n",
    "    \n",
    "    # Predicting from predictor varialbes based on the model.\n",
    "    y_test_fold_pred = lda.predict(x_test_fold)\n",
    "    \n",
    "    #store results\n",
    "    lda_accuracies_1.append(lda.score(x_test_fold, y_test_fold))\n",
    "    lda_f1scores_1.append(metrics.f1_score(y_test_fold, y_test_fold_pred, average='macro'))\n",
    "    lda_precision_1.append(metrics.precision_score(y_test_fold, y_test_fold_pred, average='macro'))\n",
    "    lda_recall_1.append(metrics.recall_score(y_test_fold, y_test_fold_pred, average='macro'))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.3951612903225806, 0.49357405140758875, 0.5060700169395822, 0.5207336523125997, 0.5552244039270687]\n",
      "[0.5, 0.49151515151515157, 0.5087755102040816, 0.5364145658263306, 0.5564516129032258]\n",
      "[0.44144144144144143, 0.3721312553429342, 0.4138071895424837, 0.47435897435897434, 0.414595452141724]\n"
     ]
    }
   ],
   "source": [
    "print(lda_precision_1)\n",
    "print(lda_recall_1)\n",
    "print(lda_f1scores_1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison of F1 Scores\n",
    "\n",
    "The closer the score is to 1, the better the classifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logistic Regression average F1 score is  0.44254308653965735\n",
      "Naive Bayes average F1 score is  0.4837996960812186\n",
      "LDA average F1 score is  0.4232668625655115\n"
     ]
    }
   ],
   "source": [
    "def average(n):\n",
    "    sum = 0\n",
    "    for i in n:\n",
    "        sum = sum + i         \n",
    "    return sum / len(n)\n",
    "\n",
    "print(\"Logistic Regression average F1 score is \", average(logit_f1scores_1))\n",
    "print(\"Naive Bayes average F1 score is \", average(naive_f1scores_1))\n",
    "print(\"LDA average F1 score is \", average(lda_f1scores_1))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "                Naive Bayes scored the highest. Yeehaw!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
